tensorflow/core/util/mkl_threadpool.h

```diff
@@ -89,18 +91,44 @@ struct MklDnnThreadPool : public threadpool_iface {
     int nthr = get_num_threads();
     int njobs = std::min(n, nthr);
     bool balance = (nthr < n);
-    for (int i = 0; i < njobs; i++) {
-      eigen_interface_->ScheduleWithHint(
-          [balance, i, n, njobs, fn]() {
-            if (balance) {
-              int start, end;
-              balance211(n, njobs, i, &start, &end);
-              for (int j = start; j < end; j++) fn(j, n);
-            } else {
-              fn(i, n);
-            }
-          },
-          i, i + 1);
+
+    if (ThreadPoolUseCallerThread() && nthr == port::NumSchedulableCPUs()) {
+      // schedule njobs-1 jobs to thread pool
+      for (int i = 0; i < njobs - 1; i++) {
+        eigen_interface_->ScheduleWithHint(
+            [balance, i, n, njobs, fn]() {
+              if (balance) {
+                int start, end;
+                balance211(n, njobs, i, &start, &end);
+                for (int j = start; j < end; j++) fn(j, n);
+              } else {
+                fn(i, n);
+              }
+            },
+            i, i + 1);
+      }
+      // run last job in caller thread
+      if (balance) {
+        int start, end;
+        balance211(n, njobs, njobs - 1, &start, &end);
+        for (int j = start; j < end; j++) fn(j, n);
+      } else {
+        fn(n - 1, n);
+      }
+    } else {
+      for (int i = 0; i < njobs; i++) {
+        eigen_interface_->ScheduleWithHint(
+            [balance, i, n, njobs, fn]() {
+              if (balance) {
+                int start, end;
+                balance211(n, njobs, i, &start, &end);
+                for (int j = start; j < end; j++) fn(j, n);
+              } else {
+                fn(i, n);
+              }
+            },
+            i, i + 1);
+      }
     }
```

Comment:
We can be more concise. Let's refactor the code in the lambda function and condition the loop bound.

Refactor:
```c++
inline void RunJobs(bool balance, int i, ...) {
  if (balance) {
    int start, end;
    balance211(n, njobs, i, &start, &end);
    for (int j = start; j < end; j++) fn(j, n);
  } else {
    fn(i, n);
  }   
}
```
The code body:
```c++
    const bool use_caller_thread =
        ThreadPoolUseCallerThread() && nthr == port::NumSchedulableCPUs();
    // Schedule njobs-1 jobs to the thread pool
    const int njobs_to_schedule = use_caller_thread ? njobs - 1 : njobs;
    for (int i = 0; i < njobs_to_schedule; i++) {
      eigen_interface_->ScheduleWithHint(
          [balance, i, n, njobs, fn]() { RunJobs(balance, i, n, njobs, fn); },
          i, i + 1);
    }
    if (use_caller_thread) {
      RunJobs(balance, njobs - 1, n, njobs, fn);
    }
```

Reply:
I have refactored the code.